code/modules/paperwork/contract.dm

```diff
@@ -0,0 +1,227 @@
+
+
+/obj/item/weapon/paper/contract/infernal/attackby(obj/item/weapon/P, mob/living/carbon/human/user, params)
+	add_fingerprint(user)
+	if(istype(P, /obj/item/weapon/pen) || istype(P, /obj/item/toy/crayon))
+		if(user.IsAdvancedToolUser())
+			if(user.mind == target)
+				if(user.mind.soulOwner == user.mind)
+					if (contractType == CONTRACT_REVIVE)
+						user << "<span class='notice'>You are already alive, this contract would do nothing.</span>"
+						return
+					else
+						user << "<span class='notice'>You quickly scrawl your name on the contract</span>"
+						if(FulfillContract()<=0)
+							user << "<span class='notice'>But it seemed to have no effect, perhaps even Hell itself cannot grant this boon?</span>"
+						return
+				else
+					user << "<span class='notice'>You are not in possession of your soul, you may not sell it.</span>"
+					return
+			else
+				user << "<span class='notice'>Your signature simply slides off of the sheet, it seems this contract is not meant for you to sign.</span>"
+				return
+		else
+			user << "<span class='notice'>You don't know how to read or write.</span>"
+			return
+	else if(istype(P, /obj/item/weapon/stamp))
+		user << "<span class='notice'>You stamp the paper with your rubber stamp, however the ink ignites as you release the stamp.</span>"
+		burn_timer += 1
+		return
+	else if(P.is_hot())
+		user.visible_message("<span class='danger'>[user] brings [P] next to [src], but [src] does not catch fire!</span>", "<span class='danger'>The [src] refuses to ignite!</span>")
+
+/obj/item/weapon/paper/contract/infernal/attack(mob/M, mob/living/user)
+	if (contractType == CONTRACT_REVIVE && target == M && M.stat == DEAD && M.mind.soulOwner == M.mind)
+		var/mob/living/carbon/human/H = M
+		var/mob/dead/observer/ghost = H.get_ghost()
+		if(ghost)
+			ghost.notify_cloning("A demon has offered you revival, at the cost of your soul.",'sound/effects/genetics.ogg', H)
+			var/response = tgalert(ghost, "A demon is offering you another chance at life, at the price of your soul, do you accept?", "Infernal Resurrection", "Yes", "No", "Never for this round", 0, 200)
+			if(!ghost)
+				return		//handle logouts that happen whilst the alert is waiting for a response.
+			if(response == "Yes")
+				H.revive()
+				add_logs(user, H, "demonically revived")
+				user.visible_message("<span class='notice'>With a sudden blaze, [H] stands back up.</span>")
+				H.adjust_fire_stacks(20)
+				FulfillContract(H)
+				sleep(10)
+				H.ExtinguishMob()
+				H.adjustFireLoss(0)
+	else
+		..()
+
+
+/obj/item/weapon/paper/contract/infernal/proc/FulfillContract(mob/living/carbon/human/user = target.current)
+	. = 1
+	world << "MEEP:  Signing contract now, user is [user]"
+	world << "MEEP: Signing contract, user type is [user.type]"
+	signed = 1
+	burn_timer += 10
+	update_text(user)
+	switch(contractType)
```

Comment:
/obj/item/weapon/paper/contract/infernal/power  seems like an awfully long typestring, but okay, I'll change that.